code/data_structures/src/heaps/binomial_heap/binomial_heap.hpp

```diff
@@ -0,0 +1,609 @@
+	 * 		Consolidating two heaps maintains merged ordering
+	 *
+	 * 	HeapA - root of heap to union with heapB
+	 * 	HeapB - root of heap to union with heapA
+	 *
+	 * 	returns
+	 * 		the root of the new heap that contains all heapA and heapB nodes
+	 *
+	 */
+	static binomialNode<T>* mergeHeap(binomialNode<T>* heapA, binomialNode<T>* heapB);
+	/*
+	 * order
+	 * 		Number of children incident on heap
+	 * 	heap
+	 * 		pointer to root of binomial heap
+	 * 	returns
+	 * 		order or heap
+	*/
+	static int order(binomialNode<T>* heap);
+private:
+	/*
+	 * head
+	 * 		pointer to the head of the heap list where every
+	 * 		node is the root node of a heap
+	 */
+	binomialNode<T> * head;
+};
+
+
+template <typename T>
+binomialHeap<T>::binomialHeap()
+{
+	head = 0;
+}
+
+template <typename T>
+binomialHeap<T>::~binomialHeap()
+{
+	if (head)
+		delete head;
+}
+
+template <typename T>
+bool binomialHeap<T>::isEmpty()
+{
+	return (head==0);
+}
+
+
+template <typename T>
+void binomialHeap<T>::insert(T v)
+{
+	binomialNode<T>*  p = head?  head->parent:  0;
+	binomialNode<T> * singleton = new binomialNode<T>(v, p);
+	head = unionHeap(singleton, head);
+
+}
+
+template <typename T>
+T binomialHeap<T>::getMin()
+{
+	T min;
+	if(head)
+	{
+		binomialNode<T> * temp = head;
+		min = head->value;
+		while(temp)
+		{
+			if(temp->value < min)
+			{
+				min = temp->value;
+			}
+			temp = temp->sibling;
+		}
+
+	}
+	return min;
+}
+template <typename T>
+T binomialHeap<T>::extractMin()
+{
+	binomialNode<T>*prev=0, *min=0, *temp=0, *next=0;
+	T minValue;
+	if(head)
+	{
+		binomialNode<T>* minPrev = 0; //previous node of min node
+		min = head;
+		temp = head->sibling;
+		prev= head;
+		/*
+		 * Find min and minPrev of heaps
+		 */
+		while(temp)
+		{
+			if(temp->value < min->value)
+			{
+				min = temp;
+				minPrev= prev;
+			}
+			prev= prev->sibling;
+			temp = temp->sibling;
+		}
+		/*
+		 *  If prev, assign prev sibling to min sibling.
+		 *  Else if no prev min->sibling is new head
+		 */
+		if(minPrev)
+		{
+			minPrev->sibling=min->sibling;
+		}
+		else
+		{
+			head = min->sibling;
+		}
+		/*
+		 * Update all children nodes parent pointers to null
+		*/
+		next = min->child;
+		temp = next;
+		while(temp)
+		{
+			temp->parent = min->parent;
+			temp = temp->sibling;
+		}
+		/*
+		 * Detach and delete min
+		 */
+		min->sibling = 0;
+		min->child = 0;
+		min->parent = 0;
+		minValue=  min->value;
+		delete min;
+		/*
+		 * Union the two detached heaps
+		 */
+		head = unionHeap(head, next);
+		temp = min = minPrev = next= 0;
+	}
+	else
+		std::cout << "\n\n ERROR couldn't not remove min!";
+
+	return minValue;
+}
+template <typename T>
+void binomialHeap<T>::print(std::ostream& out)
+{
+	if(head)
+	{
+		head->print(out);
+	}
+
+}
+/*
+ *
+ */
+template <typename T>
+int binomialHeap<T>::size()
+{
+	int i = 0;
+	if(head)
+		head->size(i);
+	return i;
+}
+
+template <typename T>
+bool binomialHeap<T>::deleteKey(T v)
```

Comment:
Instead of returning a boolean, instead just throw an exception is something goes wrong.